### PR TITLE
[Simulator] Add high-fidelity CPU-based inference simulator

### PR DESCRIPTION
**Pull request by littlefatfat**
## Motivation

This PR introduces SGLang Simulator, a CPU-based inference simulator for
evaluating SGLang scheduling, latency, throughput, and prefix-cache behavior
without loading model weights or executing model kernels.

It is a current-main rework of the simulator prototype in
[#22250](https://github.com/sgl-project/sglang/pull/22250), while retaining
SGLang v0.5.16 as the tested compatibility floor.

Large-scale inference configuration studies are expensive to run directly on
accelerators. SGLang Simulator reuses the real SGLang scheduler, radix cache,
hierarchical cache, and request lifecycle, while replacing model-forward
execution with a latency predictor. It supports measured traces, synthetic
workloads, and both in-process and HTTP serving workflows.

## What This PR Adds

- Add `tools/sglang-simulator` as a monorepo component.
- Reuse the SGLang scheduler, cache allocator, radix cache, hierarchical cache,
  and request lifecycle with dummy model weights and a minimal KV allocation.
- Add three forward-latency predictors:
  - AIConfigurator;
  - an sklearn-compatible ML predictor;
  - batch-composition replay.
- Support Random, ShareGPT, and timestamped Autobench workloads.
- Support direct Python execution through `SGLangBenchmarkRunner`.
- Support server and benchmark-client execution through the HTTP serving path.
- Collect request, latency, throughput, prefix-cache, cache-tier, and predictor
  metrics.
- Add reusable example configurations, workloads, tokenizer assets, and replay
  fixtures.
- Add compatibility checks for SGLang launch and serving APIs.
- Run the simulator regression suite from CPU PR CI when
  `tools/sglang-simulator` changes.

## Simulation Modes

The simulator provides two execution modes:

- `OFFLINE`: advances a logical clock without sleeping, intended for fast
  configuration and workload studies.
- `BLOCKING`: sleeps for predicted forward latency and visible L2-to-L1 load
  latency, intended for serving-path and timing-semantics validation.

Replay replaces model-forward latency only. CPU preprocessing and
postprocessing overhead remain enabled by default.

For `OFFLINE` runs, latency, throughput, and duration should be read from the
server-side simulator metrics rather than benchmark-client wall-clock time.

## Accuracy Validation

The following results compare simulator output with measured serving traces.
Latency, throughput, and duration values are absolute percentage errors.
Prefix reuse is reported as absolute percentage-point error.

The `SGLang baseline` column identifies the version used by the real service
that produced the measurement trace.

| Model | SGLang baseline | Hardware | Predictor / coverage | TTFT | TPOT | ITL | Input throughput | Duration | Prefix reuse |
|:---|:---:|:---:|:---|---:|---:|---:|---:|---:|---:|
| Qwen3-8B | v0.5.9 | H20 | AIConfigurator; no-cache, L1, and L2 cases | 2.38–4.15% | 1.64–4.05% | 1.78–4.07% | 1.41–2.40% | 1.39–2.33% | 0.00–0.04 pp |
| Qwen3-32B-FP8 | v0.5.9 | H20 | AIConfigurator; no-cache, L1, and L2 cases | 2.40–3.05% | 0.58–1.11% | 0.52–1.02% | 0.52–1.13% | 0.51–1.12% | 0.00–0.04 pp |
| GLM-5.1-FP8 | v0.5.11 | GPU | ML; L1 and L2 cases | 5.27% | N/A | N/A | 1.11% | 1.08% | 0.028 pp |
| DeepSeek-V4-Pro | v0.5.11 | GPU | ML; L1 and L2 cases | 6.12% | N/A | N/A | 0.23% | 0.22% | 0.048 pp |
| DeepSeek-V4-Flash | v0.5.16 | GPU | ML; L1 case | 0.786–1.343% | N/A | N/A | 0.000212% | 0.000212% | 0.000 pp |
| DeepSeek-V4-Pro | v0.5.16 | GPU | AIConfigurator; L2, 32–128K traces | 1.99–10.17% | N/A | N/A | 0.0055–0.0183% | 0.0055–0.0183% | 0.000–0.011 pp |

TPOT and ITL are not applicable to the GLM-5.1-FP8,
DeepSeek-V4-Pro, and DeepSeek-V4-Flash validation traces because they use
one-token outputs(Prefill-Decode Disaggregation).

aic-accuracy-32-128k-pr

The SGLang v0.5.16 validation additionally covers four measured 32–128K
serving traces. Using AIConfigurator as the forward-latency predictor, the
simulator achieves a mean-TTFT MAPE of 5.89%. Duration and input-throughput
MAPE are both 0.011%, while the maximum absolute prefix-reuse error remains
below 0.011 percentage points.

Prefix-cache accuracy is particularly sensitive to `max_total_tokens`. This
value determines simulated device KV-cache capacity and also affects
hierarchical host-cache sizing and eviction behavior. For deployment-faithful
comparisons, use the `max_total_num_tokens` value reported by the real SGLang
server startup log as the simulator's `--max-total-tokens` value.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #33785893262](https://github.com/sgl-project/sglang/actions/runs/33785893262)
Latest PR Test (Extra): :no_entry_sign: [Run #33830599147](https://github.com/sgl-project/sglang/actions/runs/33830599147)
Latest PR Test (AMD ROCm 7.2): :heavy_minus_sign: **No AMD PR run found for this commit**.

### CODE REVIEW

**Comment by hzh0425** [comment]
should fix the lint

File: tools/sglang-simulator/src/sglang_simulator/simulation/sglang/model_runner.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+from sglang_simulator.hook import BaseHook
+from sglang_simulator.simulation.manager import ConfigManager
+from sglang_simulator.simulation.sglang.utils import (
+    resolve_model_info,
+    resolve_scheduler_config,
+)
+from sglang_simulator.simulation.utils import profile_device_available_bytes
+
+
+class _MockModel(torch.nn.Module):
```
Comment (sagearc) [inline tools/sglang-simulator/src/sglang_simulator/simulation/sglang/model_runner.py:11]: Can we load the actual model on the [meta device](https://docs.pytorch.org/docs/2.13/meta.html) instead of using an empty mock, so we reuse its attention metadata and sliding-window behavior without materializing model parameters?

Comment (littlefatfat) [reply tools/sglang-simulator/src/sglang_simulator/simulation/sglang/model_runner.py:11]: Thanks for the review—this is a helpful suggestion. We initially used an empty mock because model construction appeared to be the most GPU-dependent part of the runtime. We agree that a meta-device model could preserve the real attention metadata and model-specific sliding-window behavior without materializing weights. We’ll prototype this with our currently supported models and evaluate a meta-first approach with an explicit fallback for unsupported models or custom layers.

File: .github/workflows/_pr-test-check-changes.yml [hunk 2]
```diff
@@ -94,6 +94,10 @@ jobs:
               - "test/**/!(*.md)"
               - "rust/**"
               - "proto/sglang/runtime/v1/sglang.proto"
+            simulator:
+              - ".github/workflows/pr-test-extra.yml"
```
Comment (hzh0425) [inline .github/workflows/_pr-test-check-changes.yml:101]: > Could we export the `simulator` filter from `_pr-test-check-changes.yml`? Otherwise, simulator-only changes won’t trigger this job because `inputs.check_changes.simulator` is missing. This PR only runs it because the workflow change also matches `main_package`.

File: tools/sglang-simulator/src/sglang_simulator/simulation/sglang/bench_runner.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+from transformers import AutoTokenizer  # noqa
+
+# The sglang must be imported after the hook installer
+from sglang.srt.entrypoints.engine import Engine  # noqa
+from sglang.srt.server_args import ServerArgs  # noqa
+
+logger = get_logger("sglang_simulator")
+
+
+class SGLangSimulationEngine(Engine):
```
Comment (hzh0425) [inline tools/sglang-simulator/src/sglang_simulator/simulation/sglang/bench_runner.py]: Bench runner should be put in /Users/hunangzhangheng/home/docker-mnt/sparse/worktree/hybrid_radix_tree/benchmark

?

File: tools/sglang-simulator/README.md
Comment (ShangmingCai) [inline tools/sglang-simulator/README.md:1]: Maybe also put this in doc?

File: .github/workflows/_pr-test-check-changes.yml (hunk 2 above)
Comment (littlefatfat) [reply .github/workflows/_pr-test-check-changes.yml:101]: Fixed — `_pr-test-check-changes.yml` now exports
`simulator` from both the `run` job and the reusable workflow. Simulator-only
changes under `benchmark/simulator/**` or `tools/sglang-simulator/**` can now
correctly trigger the simulator CPU job.

File: tools/sglang-simulator/README.md
Comment (littlefatfat) [reply tools/sglang-simulator/README.md:1]: Done — I added a user-facing guide at
`docs/docs/advanced_features/sglang_simulator.mdx`. The guide covers
installation, serving, workloads, simulation modes, metrics, and the currently
supported scope.

File: tools/sglang-simulator/src/sglang_simulator/simulation/sglang/bench_runner.py (hunk 3 above)
Comment (littlefatfat) [reply tools/sglang-simulator/src/sglang_simulator/simulation/sglang/bench_runner.py]: Fixed — the benchmark runner now lives at
`benchmark/simulator/bench_runner.py`. I also updated the related imports,
tests, documentation, and serving benchmark integration to use the new
repository-level location.

**Review by hzh0425 — APPROVED** [review]
Looks good

**Review by ShangmingCai — APPROVED** [review]
I think this is a nice tool for testing and debugging. Since most code are placed in a separated tool directory, I think it won't cause any regression and is safe to merge.

File: tools/sglang-simulator/setup.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+
+setup(
+    name="sglang-simulator",
+    version=get_version(),
+    url="https://github.com/sgl-project/sglang.git",
+    description="A High-Fidelity LLM inference simulator for SGLang",
+    packages=find_packages(where="src"),
+    package_dir={"": "src"},
+    py_modules=["usercustomize"],
```
Comment (ispobock) [inline tools/sglang-simulator/setup.py:20]: `usercustomize` runs at startup in every python process in the env. With the PYTHONPATH and `SGLANG_USE_CPU_ENGINE=1` from the docs, a real sglang in the same shell also gets `get_device_capability` stubbed to (10, 0). Is the spawn ordering the reason this has to be `usercustomize`?

File: tools/sglang-simulator/src/sglang_simulator/hook/class_hook_entry.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                        "(Regex is enabled, which might cause unexpected behavior.)"
+                        if hook.REGEX
+                        else ""
+                    )
+                )
+                target_class = _builtins_build_class_(func, name, *bases, **kwargs)
+                try:
+                    hook.hook(target_class)
+                except Exception as e:
+                    logger.warning(f"Failed to hook class [{name}]. Error: {e}")
```
Comment (ispobock) [inline tools/sglang-simulator/src/sglang_simulator/hook/class_hook_entry.py]: Why warn instead of raise? A failed hook is silent here, and nothing checks that a hook ever matched.

File: .github/workflows/_pr-test-simulator-cpu.yml [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+env:
+  SGLANG_IS_IN_CI: true
+  SKIP_PR_TEST_HEALTH_CHECK: ${{ (fromJson(inputs.caller_inputs).skip_pr_test_health_check || fromJson(inputs.caller_inputs).test_parallel_dispatch || fromJson(inputs.caller_inputs).run_all_tests) && 'true' || 'false' }}
+  PR_TEST_BYPASS_MAINTENANCE_ON_MAIN: ${{ github.ref == 'refs/heads/main' && 'true' || 'false' }}
+  USE_VENV: false
+
+jobs:
+  run:
+    name: simulator-test-cpu
+    if: fromJson(inputs.check_changes).main_package == 'true' || fromJson(inputs.check_changes).simulator == 'true'
```
Comment (ispobock) [inline .github/workflows/_pr-test-simulator-cpu.yml:28]: `main_package` covers `python/sglang/**`, and simulator-test-cpu is in the pr-test-extra-finish needs list, so this gates every core PR, not just simulator ones. The body says simulator change only?

**Comment by ispobock** [comment]
we also need to add a CODEOWNERS entry for `tools/sglang-simulator/`

File: tools/sglang-simulator/setup.py (hunk 4 above)
Comment (littlefatfat) [reply tools/sglang-simulator/setup.py:20]: Yes, this is required because spawned workers may import SGLang modules before the target wrapper runs. The guard now requires both SGLANG_SIMULATOR_BOOTSTRAP=1 and SGLANG_USE_CPU_ENGINE=1. The bootstrap flag is set internally and inherited by workers, so normal SGLang processes are unaffected.

File: tools/sglang-simulator/src/sglang_simulator/hook/class_hook_entry.py (hunk 5 above)
Comment (littlefatfat) [reply tools/sglang-simulator/src/sglang_simulator/hook/class_hook_entry.py]: Agreed. Hook installation failures now propagate immediately. Required hooks are validated after scheduler initialization, while version- or configuration-specific hooks may remain unmatched.

File: .github/workflows/_pr-test-simulator-cpu.yml (hunk 6 above)
Comment (littlefatfat) [reply .github/workflows/_pr-test-simulator-cpu.yml:28]: This is intentional. Since the simulator depends on internal SGLang interfaces, main_package changes also run the CPU simulator suite as a compatibility check.

**Review by ispobock — APPROVED** [review]

**Comment by hzh0425** [comment]
> we also need to add a CODEOWNERS entry for `tools/sglang-simulator/`

let me add in next pr

**Comment by hzh0425** [comment]
image
image
The CI for the simulator has finished running, and most of the tests for `basea`/`baseb` within `base` have already been executed.

This PR introduces a separate, new directory and does not modify any other code.
